src/librustc_typeck/error_codes.rs

```diff
@@ -2425,6 +2425,83 @@ struct Bar<S, T> { x: Foo<S, T> }
 ```
 "##,
 
+E0307: r##"
+This error indicates that the `self` parameter in a method has an invalid
+"reciever type".
+
+Methods take a special first parameter, of which there are three variants:
+`self`, `&self`, and `&mut self`. The type `Self` acts as an alias to the
+type of the current trait implementor, or "receiver type". Besides the
+already mentioned `Self`, `&Self` and `&mut Self` valid receiver types, the
+following are also valid, if less common: `self: Box<Self>`,
```

Comment:
```suggestion
following are also valid: `self: Box<Self>`,
```
don't waste readers attention span on what is and isn't common; instead, indicate it implicitly by putting the less common ones last.